File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar
{
    private int _battery;
    private int _speed;
    private int _batteryDrain;
    private int _metersDriven;
    
    public RemoteControlCar(int speed, int batteryDrain)
    {
        _battery = 100;
        _metersDriven = 0;
        _speed = speed;
        _batteryDrain = batteryDrain;
    }

    public bool BatteryDrained() => _battery - _batteryDrain < 0 ? true : false;

    public int DistanceDriven() => _metersDriven;

    public void Drive()
    {
        if (_battery - _batteryDrain >= 0)
        {
            _metersDriven += _speed;
            _battery -= _batteryDrain;
        }
    }

    public static RemoteControlCar Nitro() => new RemoteControlCar(50, 4);
}

class RaceTrack
{
    private int _distance;

    public RaceTrack(int distance)
    {
        _distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        while (!car.BatteryDrained())
        {
            car.Drive();
        }

        if (car.DistanceDriven() >= _distance) return true;
        return false;
    }
}

File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    private int speed;
    private int batteryDrain;
    private int battery;
    private int distanceDriven;
    
    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
        this.battery = 100;
        this.distanceDriven = 0;
    }

    public boolean batteryDrained() {
        return (this.battery < this.batteryDrain);
    }

    public int distanceDriven() {
        return this.distanceDriven;
    }

    public void drive() {
        if (this.battery < this.batteryDrain) return;
        this.distanceDriven += this.speed;
        this.battery -= this.batteryDrain;
    }

    public int getSpeed(){
        return this.speed;
    }

    public int getBattery(){
        return this.battery;
    }

    public int getBatteryDrain(){
        return this.batteryDrain;
    }
    
    public static NeedForSpeed nitro() {
        return (new NeedForSpeed(50, 4));
    }
}

class RaceTrack {
    private int distance;
    
    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        return (this.distance <= (car.getBattery()/car.getBatteryDrain())*car.getSpeed());
    }
}
